### PR TITLE
INT882  - New Fee plan manager and 1 call in BO page 

### PR DESCRIPTION
**Pull request by Francois-Gomis**

### CODE REVIEW

File: Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        $element->setComment($this->getHtmlComment($element->getValue()));
+        $this->setElement($element);
+        $html = $this->_toHtml();
+        $this->_arrayRowsCache = null;
+        // doh, the object is used as singleton!
+        return $html;
+    }
+
+    private function getHtmlComment($feePlans): string
```
Comment (syjust-alma) [inline Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php:180]: phpdoc

File: Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        $html = '<div style="font-size: x-small" data-toggle="collapse" ><p>' . __('Fees applied to each transaction :') . '</p>';
```
Comment (syjust-alma) [inline Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php:182]: just a comment (not required)

I prefer `sprintf()` to `.` concat

to discuss on a rainy day 😉

File: Gateway/Config/PaymentPlans/PaymentPlanConfig.php [hunk 3]
```diff
@@ -83,17 +84,18 @@ public static function defaultConfigForFeePlan(FeePlan $plan): array
...
             'deferredDays' => intval($plan->deferred_days),
             'deferredMonths' => intval($plan->deferred_months),
 
-            'deferredTriggerEnable' => isset($deferred_trigger_limit_days),
-            'deferredTriggerDays' => $deferred_trigger_limit_days,
+            'deferredTriggerEnable' => !empty($deferred_trigger_limit_days),
+            'deferredTriggerDays' => intval($deferred_trigger_limit_days),
 
-            'enabled' => $plan->installments_count === 3,
+            'enabled' => 0,
```
Comment (syjust-alma) [inline Gateway/Config/PaymentPlans/PaymentPlanConfig.php:98]: By default, all plans are disabled now ?

Are you sure this is a product request ?

File: Helpers/ConfigHelper.php [hunk 4]
```diff
@@ -172,4 +192,38 @@ protected function getConfigPath($configCode): string
...
 
+    /**
+     * @param array $plans
+     *
+     * @return void
+     */
+    public function saveBasePlansConfig(array $plans): void
+    {
+        $this->saveConfig(self::BASE_PLANS_CONFIG, $this->serializer->serialize($plans), $this->storeHelper->getScope(), $this->storeHelper->getStoreId());
+        $this->saveConfig(self::BASE_PLANS_TIME, time(), $this->storeHelper->getScope(), $this->storeHelper->getStoreId());
```
Comment (syjust-alma) [inline Helpers/ConfigHelper.php:203]: Could you explain me why you save the time here ?

File: Helpers/PaymentPlansHelper.php [hunk 5]
```diff
@@ -82,28 +93,117 @@ public function paymentTriggerIsAllowed():bool
...
+        if (isset($matches[1])) {
+            $label =  __('Pay in %1 installments', $matches[1]);
+        }
+        if (isset($matches[2]) && $matches[2] != 0) {
+            $label =  __('Pay later - D+%1', $matches[2]);
+        }
+        return $label;
+    }
+
+    public function formatFeePlanConfigForBackOfficeDisplaying(FeePlan $feePlan, array $feePlanConfig = null): array
```
Comment (syjust-alma) [inline Helpers/PaymentPlansHelper.php:163]: phpdoc

File: Helpers/PaymentPlansHelper.php [hunk 6]
```diff
@@ -82,28 +93,117 @@ public function paymentTriggerIsAllowed():bool
...
+    {
+        $key = PaymentPlanConfig::keyForFeePlan($feePlan);
+        $defaultEnabled = 0;
+        if ($key === 'general:3:0:0') {
+            $defaultEnabled = 1;
```
Comment (syjust-alma) [inline Helpers/PaymentPlansHelper.php:168]: I am not comfortable when a `format` function _calculate_ a default value, this information should be stored in `$feePlanConfig` previously like for other default values)

File: Helpers/PaymentPlansHelper.php (hunk 5 above)
Comment (syjust-alma) [inline Helpers/PaymentPlansHelper.php:163]:
```suggestion
    public function formatLocalFeePlanConfig(FeePlan $feePlan, array $feePlanConfig = null): array
```
It seems you want separate `feePlan` in to concepts:
1. `fromApi`
2. `forBackOffice` (that could become `fromBackOffice`)

Maybe you could call them:
1. `apiFeePlans`
2. and `localFeePlans` ?

File: Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php (hunk 1 above)
Comment (Francois-Gomis) [reply Block/Adminhtml/System/Config/FeePlansConfigFrontModel.php]: Done

File: Gateway/Config/PaymentPlans/PaymentPlanConfig.php (hunk 3 above)
Comment (Francois-Gomis) [reply Gateway/Config/PaymentPlans/PaymentPlanConfig.php:98]: by default plan 3 installments is up .

File: Helpers/ConfigHelper.php (hunk 4 above)
Comment (Francois-Gomis) [reply Helpers/ConfigHelper.php:203]: Feature not implemented

Comment (Francois-Gomis) [reply Helpers/ConfigHelper.php:203]: remove

File: Helpers/PaymentPlansHelper.php (hunk 5 above)
Comment (Francois-Gomis) [reply Helpers/PaymentPlansHelper.php:163]: done

File: Helpers/PaymentPlansHelper.php (hunk 6 above)
Comment (Francois-Gomis) [reply Helpers/PaymentPlansHelper.php]: split

**Pull request merged**